### PR TITLE
feat(app): allow custom labware dir to be opened and reset to default

### LINKED ISSUE #4878: Add Labware: Reset labware directory
As a Run App user, I would like to be able to reset my labware directory to the default value.

Acceptance Criteria
- [ ] User able to click a link (or button) to reset source directory to default file path
- [ ] User prompted to confirm or cancel change
- [ ] Labware files are not moved

Design
https://zpl.io/VKv6O1J

### LINKED ISSUE #4879: Add Labware: Open labware directory
As a Run App user, I would like to be able to open up my labware directory in my OS’s file browser, directly from the Run App. 

Acceptance Criteria
- [ ] User able to click a link (or button) to open source folder in OS file browser

Design
https://zpl.io/VKv6O1J

### PR DESCRIPTION
**Pull request by mcous**
## overview

This PR adds CTA's to the custom labware source directory UI to allow the user to:

- Open the directory on their computer
- Change the directory to another folder
- Reset the directory to its default

Closes #4878, closes #4879 

## changelog

- feat(app): allow custom labware dir to be opened and reset to default

## review requests

ACs copied from tickets

### reset directory

- [x] User able to click a link (or button) to reset source directory to default file path
- [x] User prompted to confirm or cancel change
- [x] Labware files are not moved

### open directory

- [x] User able to click a link (or button) to open source folder in OS file browser

### CODE REVIEW

File: app/src/config/index.js [hunk 1]
```diff
@@ -26,6 +31,15 @@ export function updateConfig(path: string, value: any): UpdateConfigAction {
   }
 }
 
+// trigger a config value reset via the app-shell via shell middleware
+export function resetConfig(path: string): ResetConfigAction {
+  return {
+    type: 'config:RESET',
```
Comment (iansolano) [inline app/src/config/index.js]: similar to comment as above around having these types be consts and then importing elsewhere

Comment (iansolano) [reply app/src/config/index.js]: `RESET_CUSTOM_LABWARE_DIRECTORY`? more explicit and similar naming? I know the other configs aren't like this though. also, realizing this file isn't an `actions` file which is a little confusing (i know this is more electron specific)

File: app/src/custom-labware/__tests__/actions.test.js [hunk 2]
```diff
@@ -2,13 +2,13 @@
...
 
 type ActionSpec = {|
   name: string,
   creator: (...Array<any>) => mixed,
   args?: Array<mixed>,
-  expected: CustomLabwareAction,
+  expected: Action,
 |}
 
 describe('custom labware actions', () => {
```
Comment (iansolano) [inline app/src/custom-labware/__tests__/actions.test.js]: use constant in tests? not sure if you are intentionally avoiding doing this, curious to hear why if so?

File: app-shell/src/config.js [hunk 3]
```diff
@@ -104,17 +105,21 @@ const log = () => _log || (_log = createLogger('config'))
 // initialize and register the config module with dispatches from the UI
 export function registerConfig(dispatch: Dispatch) {
   return function handleIncomingAction(action: Action) {
-    if (action.type === 'config:UPDATE') {
-      const { payload } = action
+    if (action.type === 'config:UPDATE' || action.type === 'config:RESET') {
```
[the remark this answers is not in the document]
Comment (mcous) [reply app-shell/src/config.js]: Totally agreed, happy to include in this PR if you can review. Will ping you when pushed

File: app/src/components/AddLabwareCard/__tests__/ConfirmResetPathModal.test.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  test('renders an AlertModal', () => {
+    const wrapper = shallow(<ConfirmResetPathModalTemplate {...props} />)
+
+    expect(wrapper.exists(AlertModal)).toBe(true)
+  })
+
+  test('renders a cancel button that calls props.onCancel', () => {
+    const wrapper = mount(<ConfirmResetPathModalTemplate {...props} />)
+    const button = wrapper.find('button[name="cancel"]')
```
[the remark this answers is not in the document]
Comment (mcous) [reply app/src/components/AddLabwareCard/__tests__/ConfirmResetPathModal.test.js]: Oops, oversight. Will fix

File: app/src/config/index.js (hunk 1 above)
Comment (mcous) [reply app/src/config/index.js]: > also, realizing this file isn't an actions file which is a little confusing

It is a little confusing. The `app/src/config` module predates our modules structure recommendations a little. Will re-org with the action type constants change

File: app/src/custom-labware/__tests__/actions.test.js (hunk 2 above)
Comment (mcous) [reply app/src/custom-labware/__tests__/actions.test.js]: This is intentional. If I use the constant in the action test but accidentally fail to ever define the constant (or make a typo in the constant value), the test will pass even though the action creator is "broken".

In tests that aren't action tests (e.g. reducer tests) I am comfortable importing action type constants and action creator functions

File: app/src/config/index.js (hunk 1 above)
Comment (iansolano) [reply app/src/config/index.js]: > resetCustomLabwareDirectory action creator in the labware module that creates a config:RESET action object by calling config::resetConfig internally

yeah, i saw this - makes sense. I guess the thing here is this `reset` action _could_ potentially do a variety of things. Seems fine for now though, we can always come back to this.

File: app/src/custom-labware/__tests__/actions.test.js (hunk 2 above)
Comment (iansolano) [reply app/src/custom-labware/__tests__/actions.test.js]: i should have mentioned I meant importing from the actions file. the typo comment you make is exactly what I was thinking about here tbh with regards to abstaining / forcing an eng to also update the test here. sounds good, let's keep this pattern.

File: app/src/components/AddLabwareCard/__tests__/AddLabware.test.js [hunk 5]
```diff
@@ -15,13 +15,7 @@ describe('AddLabware', () => {
     const wrapper = mount(<AddLabware onAddLabware={mockOnAddLabware} />)
 
     expect(mockOnAddLabware).toHaveBeenCalledTimes(0)
-    wrapper.find('button').invoke('onClick')()
+    wrapper.find('button[name="add-labware"]').invoke('onClick')()
```
Comment (iansolano) [inline app/src/components/AddLabwareCard/__tests__/AddLabware.test.js]: i think you just missed this one but prob should be a const like the others

File: app/src/config/actions.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
+// @flow
```
Comment (iansolano) [inline app/src/config/actions.js]: nice!

File: app/src/components/AddLabwareCard/__tests__/AddLabware.test.js (hunk 5 above)
Comment (mcous) [reply app/src/components/AddLabwareCard/__tests__/AddLabware.test.js]: Dammit

**Pull request merged**